### ISSUE TITLE
rpm pre-install scripts rely on selinux being installed

### ISSUE THREAD

**Issue by wouterhund** [schedulix/schedulix#14]
In aa5a1b37d0a38441d02671d1c6f4c72ebf1ad69d the pre-install script was modified to include calls to `getenforce` and `setenforce`. This breaks the script when selinux is not present, which is the case in docker containers and other minimal installations.

Ideally the rpm would include selinux policies so this toggling of the enforce mode (which is a temporary system-wide change) would not be necessary. Alternatively the script should check for the presence of selinux, and only execute these commands when selinux is actually present and active.

**Comment by schedulix** [comment]
Hi Wouter,

thank you for raising this issue. TBH I'm not at all a specialist if it's about SELinux.
If you happen to know how to add those policies, feel free to explain this to me.

Detecting the presence of SELinux seems to be the easier but less sophisticated approach.
I will have a look at this. Maybe it's just as simple as 

if [ -x /sbin/getenforce ]; then blablabla; ... 

But I'm definitely open for suggestions from your side.
For longer explanations you can use e-mail (ronald dot jeninga at independit dot de) if you want

**Comment by wouterhund** [comment]
Starting with simply detecting the presence of selinux would be a good start. We're currently working around this problem by providing dummy shell scripts named `getenforce` and `setenforce` that simply do nothing.

Unfortunately I'm not very familiar with the policy writing side of selinux either, so I can't assist with that. I believe the easiest way is to run the application on a system with selinux running in permissive mode, then using `audit2allow` to convert the audit logs to an initial policy file, which can then be adjusted by hand. This process is described here: https://wiki.centos.org/HowTos/SELinux#Creating_Custom_SELinux_Policy_Modules_with_audit2allow

For reference, this is the script we're currently using to work around the dependency on `getenforce` and `setenforce` being present:

```
#!/bin/bash
set -euo pipefail

# Remove our shim
if [[ "$1" == "off" ]]; then
    rm /usr/sbin/getenforce /usr/sbin/setenforce
    exit 0
fi

# Check if getenforce is already present, if so, bail
if [[ -f /usr/sbin/getenforce ]]; then
    echo "getenforce is already present!"
    echo "check if docker image now ships with selinux and if shim-getenforce is still needed"
    exit 1
fi

cat <<EOF > /usr/sbin/getenforce
#!/bin/bash
echo 0
EOF

cat <<EOF > /usr/sbin/setenforce
#!/bin/bash
EOF

chmod +x /usr/sbin/getenforce /usr/sbin/setenforce
```

**Comment by schedulix** [comment]
Thank you. In fact the only problem to get around is the user creation from within the rpm.
We don't have any issues with SELinux elsewhere.
I think the following change would be helpful for you. That'll give me time to investigate the matter deeper:

```
        if [ -x /sbin/getenforce ]; then
                SELINUX_SETTING=`getenforce`
                setenforce 0
        fi
        if ! grep schedulix /etc/passwd >/dev/null 2>&1; then
                useradd schedulix -d /opt/schedulix -m -s /bin/bash -U -p "RJteetpJ9UFeQ";
        fi; \
        if [ ! -d /opt/schedulix ]; then
                mkdir -p /opt/schedulix;
        fi; \
        chmod 755 /opt/schedulix;
        chown schedulix.schedulix /opt/schedulix
        if [ -n "$SELINUX_SETTING" ]; then
                setenforce $SELINUX_SETTING
        fi

```
Do you use the rpms created by me, or do you create rpms yourself (make rpm does the trick)?
In case of the former I'll create and upload them (version?). In case of the latter you could test if the change repairs the issue.
On my CentOS /sbin/getenforce is hard linked to /usr/sbin/getenforce (or vice versa ;).

Best regards,

Ronald

**Comment by wouterhund** [comment]
That change would indeed be helpful, thank you!

We're using the rpms hosted at https://www.independit.de/repo/centos/7/x86_64/, currently the latest version (`2.9-16.el7`)

**Comment by schedulix** [comment]
I've just uploaded the rpms and rebuilt the meta data. You should be able to see the 2.9-17.el7 now.
Please report if the issue is resolved with the small change.

**Comment by wouterhund** [comment]
Looks good, our build passes with `2.9-17.el7` and no selinux related workarounds. Thanks!

**Comment by schedulix** [comment]
OK, I consider the issue closed then. Thank you for the response.

**Comment by rafaelorafaelo** [comment]
Hi, is it possible to reopen this issue? The problem persists.

My systems are SELinux disabled. `SELINUX_SETTING='getenforce'` sets SELINUX_SETTING to Disabled. Later you call `setenforce $SELINUX_SETTING`, but Disabled is not an allowed parameter for setenforce (see `man setenforce`). The script fails due to this wrong parameter and the RPM installation fails.

The solution would be simple: if SELinux is disabled, never call setenforce. There is also another program to check the SELinux state: see `man selinuxenabled`. It's not possible to disable SELinux with setenforce 0, it sets SELinux only to permissive mode. The only way to disable SELinux is to set a configuration variable and reboot the system.

Regards
Ralf

**Comment by schedulix** [comment]
Sure, I'm not against a more accurate treatment of selinux. I'll have a look at it and I'll fix it for both 2.9 and the upcoming 2.10.
But since I'm in the middle of doing the final work for releasing 2.10, it might take a few days.

Is that OK for you?

Regards,

Ronald

**Comment by rafaelorafaelo** [comment]
Hi Ronald, it's ok. 

I slightly modified your code to skip all SELinux commands when SELinux is disabled.

Regards,
Ralf

```sh
echo "executing pre base -- 2.9-22.el7"
if [ "$1" == "1" ]; then
	if [ -x /usr/sbin/selinuxenabled -a -x /sbin/getenforce ]; then
		if /usr/sbin/selinuxenabled; then
			SELINUX_SETTING=`getenforce`
			setenforce 0
		fi
	fi
	if ! grep schedulix /etc/passwd >/dev/null 2>&1; then
		useradd schedulix -d /opt/schedulix -m -s /bin/bash -U -p "RJteetpJ9UFeQ";
	fi; \
	if [ ! -d /opt/schedulix ]; then
		mkdir -p /opt/schedulix;
	fi; \
	chmod 755 /opt/schedulix;
	chown schedulix.schedulix /opt/schedulix
	if [ -n "$SELINUX_SETTING" ]; then
		setenforce $SELINUX_SETTING
	fi
else
	echo "Upgrading schedulix-base";
	# if there is some server installed, stop it (if it's running, but the server-stop utility is forgiving if not)
	if [ -x /etc/init.d/schedulix-server ]; then
		service schedulix-server stop || true
	fi
	if [ -x /etc/init.d/schedulix-client ]; then
		service schedulix-client stop || true
	fi
fi

```

**Comment by schedulix** [comment]
Hi Ralf,

thank you! 
I've just checked that both selinuxenabled and getenforce are part of the same package (libselinux-utils-2.5-15.el7.x86_64 : SELinux libselinux utilies). Therefor if the one is present, then the other is too. This is at least true on CentOS7/RHEL7.
I don't doubt that it'll be true for CentOS8/RHEL8 as well. (checked: libselinux-utils-2.9-4.el8_3.x86_64 : SELinux libselinux utilities)

Now if getenforce would be present but selinuxenabled wouldn't, or vice versa, the code would break. So either we'll have to cover those situations, or agree that if we have the one then we have the other and simplify the if statement.
(just as the original code assumed the presence of setenforce if getenforce exists).
Given the fact that both 7 and 8 provide both utilities, we'd have to wait for CentOS9/RHEL9 before the code possibly breaks. I'm willing to take that risk ;-)

Do you agree?

Regards,

Ronald

**Comment by rafaelorafaelo** [comment]
Hi Ronald, I also checked the presence and packages of selinuxenabled  in CentOS 7/8 before sending the code. I think there is no risk at all. The only risks: CentOS 8 is EOL 12.2021. CentOS 9 will never come. But there is a bunch of new RHEL clones on the way. Or do you prefer CentOS Stream aka RHEL rolling release?

Regards
Ralf

**Comment by schedulix** [comment]
Hi Ralf, it's all about code stability. And if the current code runs on CentOS/RHEL 7 and 8, I'm fine for now. Thinking about how long those releases live, there's nothing to worry about. In the future things will be different, no doubt. As long as there isn't a revolution every year, I'll be able to handle that.
I've simplified the if statement and I'll do a push both for 2.9 and 1.10 some time today.
As soon as I desperately need something else to do than writing documentation, I'll create the new rpms.

**Comment by schedulix** [comment]
Hi Ralf, I've now uploaded freshly built rpms for 2.9 (latest stable) and 2.10 (pre-release) for both CentOS/RHEL 7 and 8.
They should now behave as desired.
If you confirm this, I'll close the issue. And of course, if I managed to make some stupid mistake, please tell me.

**Comment by rafaelorafaelo** [comment]
Hi Ronald, I installed version 2.9.23 and 2.10.4 on CentOS 7 and a CentOS 8 box with SELinux disabled, a fresh installation. No problems. I tried 2.9.22 which stopped with the error. Looks like the problem is solved. Thanks a lot.

**Comment by schedulix** [comment]
Hi Ralf,

thank you very much for testing and verifying that we've resolved the issue.
I consider the issue as closed now.

Best regards,

Ronald

**Comment by rafaelorafaelo** [comment]
Hi Ronald, could you please fix the EL7 packages? I accidentally installed the EL8 packages in CentOS 7, which works great. But now I use the proper path https://www.independit.de/repo/schedulix-2.9/centos/7/x86_64/ which contains wrong code in schedulix-base-2.9-23.el7.x86_64.rpm.

**Comment by schedulix** [comment]
Hi Ralf,

sorry for the hassle and thank you for pointing me at this.
I must have been confused while hopping between CentOS 7 and CentOS 8 machines and keeping the git sandboxes sync'ed.
As soon as I find the time I'm going to automate the rpm building process. I know of a nice scheduling system that would be able to handle that ;-)

Anyway, I've fixed it, but note that I kept the release number (if you would have cloned the git repository and would have done a "make rpm", you would have a 2.9-23 with the correct pre-script). I've also uploaded the rpm. The next attempt should be successful.

Thanks again,

Ronald
